include/gatekeeper_gk.h

```diff
@@ -122,6 +140,13 @@ struct gk_fib {
 			 */
 			struct ether_cache eth_cache;
 		} grantor;
+
+		/*
+		 * When the action is GK_FWD_BACK_NET or GK_FWD_FRONT_NET,
+		 * we need the next fib entry for the gateway of the
+		 * destination.
+		 */
+		struct gk_fib *next_fib;
```

Comment:
Notice that if the next FIB entry is `GK_FWD_NEIGHBOR_FRONT_NET` or `GK_FWD_NEIGHBOR_BACK_NET`, we won't have a cached Ethernet header.

Instead of copying the strategy we did in grantor entries, I recommend haveing a perfect hash in the neighbor entries to cache the Ethernet header of the neighbor. This will simplify the code because it will work for LPM prefixes mapping directly to neighbor entries and for the grantor case.

The perfect hash can be easily accomplished using the length of the prefix that maps to the FIB entry. For example, the prefix 10.0.0.0/24 requires 2^(32-24) = 256 buckets to be a perfect hash with the hash function being: `IP_addr & ((1 << (32-24)) - 1)`. Of course, this bucket array must be dynamically allocated.

Moreover, each bucket must have the Ethernet header and a ref counter to count how many grantor entries are pointing to that neighbor. This way, we can eventually release a neighbor that isn't being referred by anyone.